src/database/Database.cpp

```diff
@@ -443,6 +444,44 @@ Database::selectUpdateMap(
     return vecT.size();
 }
 
+template <typename T>
+void
+Database::decodeOpaqueXDR(std::string const& in, T& out)
+{
+    std::vector<uint8_t> opaque;
+    decoder::decode_b64(in, opaque);
+    xdr::xdr_from_opaque(opaque, out);
+}
+
+template <typename T>
+void
+Database::decodeOpaqueXDR(std::string const& in, soci::indicator const& ind,
+                          T& out)
+{
+    if (ind == soci::i_ok)
+    {
+        decodeOpaqueXDR(in, out);
+    }
+}
+
+// Explicit instantiations of decodeOpaqueXDR() for clients to link with.
+template void
```

Comment:
This is the same thing as the `selectMap` functions. Publicly accessible template functions should be declared _and_ defined in a header file. Then you don't need explicit instantiations. For perspective: the template functions you wrote have either 3 line or 4 line implementations, yet most of these declarations are 4 lines! Furthermore, doing this leads to bloat: there won't be a compile error or link error if I stop using one of these functions yet it will persist in my executable.

Reply:
Done.